packages/element/src/store.ts

```diff
@@ -0,0 +1,900 @@
+    const nextAppStateSnapshot = !isObservedAppState(appState)
+      ? getObservedAppState(appState)
+      : appState;
+
+    const didAppStateChange = this.detectChangedAppState(
+      nextAppStateSnapshot,
+      options,
+    );
+
+    if (!didAppStateChange) {
+      return this.appState;
+    }
+
+    return nextAppStateSnapshot;
+  }
+
+  private maybeCreateElementsSnapshot(
+    elements: SceneElementsMap | undefined,
+    options: {
+      shouldCompareHashes: boolean;
+    } = {
+      shouldCompareHashes: false,
+    },
+  ): SceneElementsMap {
+    if (!elements) {
+      return this.elements;
+    }
+
+    const changedElements = this.detectChangedElements(elements, options);
+
+    if (!changedElements?.size) {
+      return this.elements;
+    }
+
+    const elementsSnapshot = this.createElementsSnapshot(changedElements);
+    return elementsSnapshot;
+  }
+
+  private detectChangedAppState(
+    nextObservedAppState: ObservedAppState,
+    options: {
+      shouldCompareHashes: boolean;
+    } = {
+      shouldCompareHashes: false,
+    },
+  ): boolean | undefined {
+    if (this.appState === nextObservedAppState) {
+      return;
+    }
+
+    const didAppStateChange = Delta.isRightDifferent(
+      this.appState,
+      nextObservedAppState,
+    );
+
+    if (!didAppStateChange) {
+      return;
+    }
+
+    const changedAppStateHash = hashString(
+      JSON.stringify(nextObservedAppState),
+    );
+
+    if (
+      options.shouldCompareHashes &&
+      this._lastChangedAppStateHash === changedAppStateHash
+    ) {
+      return;
+    }
+
+    this._lastChangedAppStateHash = changedAppStateHash;
+
+    return didAppStateChange;
+  }
+
+  /**
+   * Detect if there any changed elements.
+   *
+   * NOTE: we shouldn't just use `sceneVersionNonce` instead, as we need to call this before the scene updates.
+   */
+  private detectChangedElements(
+    nextElements: SceneElementsMap,
+    options: {
+      shouldCompareHashes: boolean;
+    } = {
+      shouldCompareHashes: false,
+    },
+  ) {
+    if (this.elements === nextElements) {
+      return;
+    }
+
+    const changedElements: SceneElementsMap = new Map() as SceneElementsMap;
+
+    for (const [id, prevElement] of this.elements) {
+      const nextElement = nextElements.get(id);
+
+      if (!nextElement) {
+        // element was deleted
+        changedElements.set(
+          id,
+          newElementWith(prevElement, { isDeleted: true }),
+        );
+      }
+    }
+
+    for (const [id, nextElement] of nextElements) {
+      const prevElement = this.elements.get(id);
+
+      if (
+        !prevElement || // element was added
+        prevElement.version < nextElement.version // element was updated
+      ) {
+        changedElements.set(id, nextElement);
+      }
+    }
+
+    if (!changedElements.size) {
+      return;
+    }
+
+    const changedElementsHash = hashElementsVersion(
+      Array.from(changedElements.values()),
+    );
```

Comment:
since this is on a hot path, I'd reduce allocs where we can.

Let's rewrite the `hashElementsVersion` so it can efficiently consume both arrays and maps:

```ts
export const hashElementsVersion = (elements: ElementsMapOrArray): number => {
  let hash = 5381;
  const _elements: Iterable<ExcalidrawElement> = Array.isArray(elements)
    ? elements
    : elements.values();
  for (const element of _elements) {
    hash = (hash << 5) + hash + element.versionNonce;
  }
  return hash >>> 0; // Ensure unsigned 32-bit integer
};
```